### PR TITLE
Add ZFS support

### PR DESCRIPTION
**Pull request by Almugra**
Allows to commit, extract, destroy and list ZFS snapshots.

Commands could be used like this: 
- `sudo zfs send test@snapshot | 0s zfs commit myStash --snapshot test@snapshot`
- `0s zfs extract myStash --snapshot test@snapshot | sudo zfs receive -F test/fs@snapshot`
- `0s zfs destroy myStash --snapshot test@snapshot` 
- `0s zfs ls myStash`

### CODE REVIEW

File: zerostash-files/src/snapshots.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+}
+
+impl Snapshot {
+    pub fn from_stdin(writer: AEADWriter) -> Result<Snapshot, SnapshotError> {
+        let mut buf = Vec::default();
+        std::io::stdin().read_to_end(&mut buf)?;
+
+        let mut sink = BufferedSink::new(writer);
+        sink.write_all(&buf)?;
```
Comment (rsdy) [inline zerostash-files/src/snapshots.rs]: So this may be in the terabytes range. I think a better approach for this would be to use a large-ish buffer of like 1MB or so, read that, and write it to the sink in a loop.

File: zerostash-files/src/snapshots.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        let stream = sink.finish()?;
+
+        let now = SystemTime::now();
+        let since_epoche = now.duration_since(SystemTime::UNIX_EPOCH).unwrap();
```
Comment (rsdy) [inline zerostash-files/src/snapshots.rs]: `since_epoch`

File: zerostash-files/src/snapshots.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        })
+    }
+
+    pub fn to_stdout(&self, reader: PoolRef<AEADReader>) -> Result<(), SnapshotError> {
+        let mut lock = std::io::stdout().lock();
+        let mut buf = Vec::default();
+
+        let mut stream = self.stream.open_reader(reader);
+        stream.read_to_end(&mut buf)?;
+        lock.write_all(&buf)?;
```
Comment (rsdy) [inline zerostash-files/src/snapshots.rs]: Pretty much the same thing as above. This may not practically fit in memory, so you should work with a loop and a static buffer.

File: zerostash-files/src/stash/list.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        iter(stash, globs)
+    }
+}
+
+pub type SnapshotIterator<'a> = Box<(dyn Iterator<Item = (String, DateTime<Utc>)> + Send + 'a)>;
+
+fn iter<V: AsRef<[T]>, T: AsRef<str>>(stash: &Infinitree<Files>, glob: V) -> SnapshotIterator {
+    let matchers = glob
+        .as_ref()
+        .iter()
```
Comment (rsdy) [inline zerostash-files/src/stash/list.rs]: Taking a reference to an `V: Iterator<Item = T>` would be more elegant

File: zerostash-files/src/stash/list.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            .iter(stash.index().snapshots(), move |snapname| {
+                if matchers.iter().any(|m| m.matches(snapname)) {
+                    Take
+                } else {
+                    Skip
+                }
+            })
+            .unwrap()
+            .filter(|(_, snap)| snap.is_some())
+            .map(|(name, snap)| {
```
Comment (rsdy) [inline zerostash-files/src/stash/list.rs]: You can use `filter_map(|(_, snap)| snap.ok()).map(...)` so you can get rid of the unwrap completely.

File: zerostash/src/commands.rs [hunk 6]
```diff
@@ -45,6 +47,27 @@ pub enum ZerostashCmd {
...
+    Commit(ZfsCommit),
+
+    /// Extracts a snapshot to stdout
+    Extract(ZfsExtract),
+
+    /// Remove a snapshot from the stash
+    Destroy(ZfsDestroy),
+
+    /// List Snapshots in a stash
+    Ls(ZfsLs),
```
Comment (rsdy) [inline zerostash/src/commands.rs]: Please move these into the `zfs` module.

File: zerostash/src/commands.rs [hunk 7]
```diff
@@ -142,6 +165,12 @@ impl Runnable for EntryPoint {
                 Ls(cmd) => cmd.run().await,
                 Keys(cmd) => cmd.run().await,
                 Wipe(cmd) => cmd.run().await,
+                Zfs(zfs) => match zfs {
+                    ZerostashZfs::Commit(cmd) => cmd.run().await,
+                    ZerostashZfs::Extract(cmd) => cmd.run().await,
+                    ZerostashZfs::Destroy(cmd) => cmd.run().await,
+                    ZerostashZfs::Ls(cmd) => cmd.run().await,
+                },
```
Comment (rsdy) [inline zerostash/src/commands.rs]: This logic should probably go to the zfs module, too.

File: zerostash/src/config/symmetric_key.rs [hunk 8]
```diff
@@ -77,7 +77,7 @@ impl SymmetricKey {
     pub fn interactive_credentials(self, stash: &str) -> Result<(SecretString, SecretString)> {
         let user = match self.user {
             Some(ref u) => u.clone(),
-            None => rprompt::prompt_reply_stderr("Username: ")?.into(),
+            None => rprompt::prompt_reply("Username: ")?.into(),
```
Comment (rsdy) [inline zerostash/src/config/symmetric_key.rs]: Revert this change, please. It's best to handle dependency upgrades separately, as this is already conflicty.

File: zerostash/src/commands/zfs/zfs_commit.rs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+fn add_snapshot(stash: &Infinitree<Files>, snapshot: String) {
+    let writer = stash.storage_writer().unwrap();
+    let stream = Snapshot::from_stdin(writer).expect("Failed to capture Snapshot");
+
+    let snapshots = &stash.index().snapshots;
+    if snapshots
+        .update_with(snapshot.clone(), |_v| stream.clone())
+        .is_none()
```
Comment (rsdy) [inline zerostash/src/commands/zfs/zfs_commit.rs]: Don't allow updating existing snapshots! This can result in accidentally overwriting data, which is not great for a backup 🙂. Let the user decide how to resolve this, just exit with a non-zero exit code here.

File: zerostash/src/commands/zfs/zfs_commit.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+impl AsyncRunnable for ZfsCommit {
+    /// Start the application.
+    async fn run(&self) {
+        let mut stash = self.stash.open();
+        stash.load(stash.index().snapshots()).unwrap();
+
+        add_snapshot(&stash, self.snapshot.clone());
+
+        stash
+            .commit(self.message.clone())
```
Comment (rsdy) [inline zerostash/src/commands/zfs/zfs_commit.rs:53]: You can use `Option::take` to avoid cloning.

File: zerostash/Cargo.toml [hunk 11]
```diff
@@ -18,7 +18,7 @@ infinitree = { version = "0.9.0", features = ["cryptobox", "yubikey"] }
 infinitree-backends = { version = "0.9.0", default-features = false, features = ["rustls"] }
 zerostash-files = { version = "0.5.0", path = "../zerostash-files" }
 rpassword = "7.0.0"
-rprompt = "1.0.5"
+rprompt = "2.0.2"
```
Comment (rsdy) [inline zerostash/Cargo.toml:21]: Please revert this!

**Review by rsdy — CHANGES_REQUESTED** [review]
The command implementations are super clean. The structural comments are mostly nitpicking, but to stay consistent with other submodules is helpful, as there's quite a lot of complexity here.

However, with large snapshots this will fall apart. You must use a deterministic amount of memory, that's independent from your input size.

File: zerostash/src/commands/zfs/zfs_commit.rs (hunk 10 above)
Comment (Almugra) [reply zerostash/src/commands/zfs/zfs_commit.rs:53]: Is it ok to change to `&mut self`? It causes a lot of errors I would have to fix.

Comment (rsdy) [reply zerostash/src/commands/zfs/zfs_commit.rs:53]: Ah, yes. It would probably make things simpler.

Comment (Almugra) [reply zerostash/src/commands/zfs/zfs_commit.rs:53]: I think because `Runnable` uses `&self` I would need to clone cmd, but that doesn't sound like its worth it to me. Am I missing something?

Comment (rsdy) [reply zerostash/src/commands/zfs/zfs_commit.rs:53]: Ah. I see. It sounds like cloning here makes sense indeed.

File: zerostash/src/commands/zfs/zfs_commands.rs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    Extract(ZfsExtract),
+
+    /// Remove a snapshot from the stash
+    Destroy(ZfsDestroy),
+
+    /// List Snapshots in a stash
+    Ls(ZfsLs),
+}
+
+pub fn match_zfs_cmd<'a>(cmd: &'a ZerostashZfs) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
```
Comment (rsdy) [inline zerostash/src/commands/zfs/zfs_commands.rs]: I think you should be able to implement `AsyncRunnable` here, just like how the other commands do it. Take a look at `keys.rs` which does this ad absurdum.

**Review by rsdy — COMMENTED** [review]
Based on the code, this looks fine. I'll do some stress tests to see if tuning can be better, but otherwise happy to merge it once you do this 1 niggle and resolve conflicts.

**Comment by Almugra** [comment]
I think the conflicts should be resolved when the master branch does a `cargo update`.

I've changed the ZFS commands to work internally so piping is no longer necessary.
The commands could be used like this:

commit: `0s zfs commit myStash -n test@first -- test@snapshot`
extract: `0s zfs extract myStash -n test@first -- -F test/fs@snapshot`

**Review by rsdy — APPROVED** [review]
